dbms/src/Common/CachingAllocator.h

```diff
@@ -0,0 +1,875 @@
+
+#include <Common/Exception.h>
+#include <Common/randomSeed.h>
+#include <Common/formatReadable.h>
+
+/// Old Darwin builds lack definition of MAP_ANONYMOUS
+#ifndef MAP_ANONYMOUS
+#define MAP_ANONYMOUS MAP_ANON
+#endif
+
+namespace DB::ErrorCodes
+{
+extern const int CANNOT_ALLOCATE_MEMORY;
+extern const int CANNOT_MUNMAP;
+}
+
+struct CachingAllocatorStats
+{
+    size_t total_chunks_size {0};
+    size_t total_allocated_size {0};
+    size_t total_size_currently_initialized {0};
+    size_t total_size_in_use {0};
+
+    size_t num_chunks {0};
+    size_t num_regions {0};
+    size_t num_free_regions {0};
+    size_t num_regions_in_use {0};
+    size_t num_keyed_regions {0};
+
+    size_t hits {0};
+    size_t concurrent_hits {0};
+    size_t misses {0};
+
+    size_t allocations {0};
+    size_t allocated_bytes {0};
+    size_t evictions {0};
+    size_t evicted_bytes {0};
+    size_t secondary_evictions {0};
+};
+
+/**
+ * Thread-safe caching allocator interface.
+ * Uses mmap as an [de]allocator tool.
+ *
+ *
+ * Motivation:
+ * <ul>
+ * <li>Cache has a specific memory usage limit (and we want it to include allocator fragmentation overhead).</li>
+ * <li>Memory fragmentation can be bridged over using cache eviction.</li>
+ * <li>There is no sense for reclaimed memory regions to be cached internally by usual allocator (malloc/new)
+ *   as we implement our own routines.</li>
+ * <li>Direct memory allocation serves for debugging purposes: placing mmap'd memory far from malloc'd facilitates
+ *   revealing memory stomping bugs.</li>
+ * </ul>
+ *
+ * @note Cache is not NUMA friendly.
+ *
+ * Performance: few million ops/sec from single thread, less in case of concurrency.
+ * Fragmentation is usually less than 10%.
+ */
+template <typename TKey, typename TMapped, typename THash = std::hash<TKey>>
+class CachingAllocator : private boost::noncopyable
+{
+private:
+    /**
+     * We will allocate memory in chunks of at least that size.
+     * 64 MB makes mmap overhead comparable to memory throughput.
+     */
+    static constexpr size_t min_chunk_size = 64 * 1024 * 1024;
+
+    static constexpr size_t page_size = 4096;
+
+    /// Sizes and addresses of allocated memory will be aligned to this boundary.
+    static constexpr size_t ptrs_alignment = 16;
+
+    struct MemoryRegion;
+
+public:
+    using Key    = TKey;
+    using Mapped = TMapped;
+    using Hash   = THash;
+
+    using MemoryRegionPtr = std::shared_ptr<MemoryRegion>;
+
+    explicit CachingAllocator(size_t max_total_size_) noexcept: max_total_size(max_total_size_) {}
+
+    ~CachingAllocator() noexcept
+    {
+        std::lock_guard cache_lock(mutex);
+
+        allocated_regions.clear();
+        unused_allocated_regions.clear();
+        free_regions.clear();
+
+        all_regions.clear_and_dispose(region_metadata_disposer);
+    }
+
+    inline size_t getSizeInUse() const
+    {
+        std::lock_guard lock(mutex);
+        return total_size_in_use;
+    }
```

Comment:
Not sure if `MarkCache`  should be given `total_size_in_use` (maybe, the caller wants to know about total allocated size or so).